### PR TITLE
BT serial: SSP improvements, added missing events

### PR DESCRIPTION
**Pull request by PilnyTomas**
Added missing events
Added method to remove cached devices.
Renamed, clarified, and extended usage of SSP API + example
Fixed support for legacy pairing (fixed PIN) - must use as IDF component and disable `CONFIG_BT_SSP_ENABLED`
related https://github.com/espressif/arduino-esp32/issues/4566

Including fix for #8448

### CODE REVIEW

File: libraries/BluetoothSerial/examples/SerialToSerialBT_SSP/SerialToSerialBT_SSP.ino [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+void serial_response(){
+  if (Serial.available()){
+      SerialBT.write(Serial.read());
+    }
+    if (SerialBT.available()){
+      Serial.write(SerialBT.read());
+    }
+    delay(20);
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/examples/SerialToSerialBT_SSP/SerialToSerialBT_SSP.ino:106]:
```suggestion
    SerialBT.write(Serial.read());
  }
  if (SerialBT.available()){
    Serial.write(SerialBT.read());
  }
  delay(20);
```

File: libraries/BluetoothSerial/examples/SerialToSerialBT_SSP/SerialToSerialBT_SSP.ino [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    Serial.println("Authenticate pairing on the other device. No PIN is used");
+  }else if(INPUT_CAPABILITY and not OUTPUT_CAPABILITY){
+    Serial.println("After pairing is initiated you will be required to enter the passkey to the ESP32 device to authenticate\n > The Passkey will displayed on the other device");
+  }
+}
+
+void loop(){
+  if (confirmRequestDone){
+    serial_response();
+  }else{
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/examples/SerialToSerialBT_SSP/SerialToSerialBT_SSP.ino:132]:
```suggestion
  } else {
```

File: libraries/BluetoothSerial/examples/bt_remove_paired_devices/bt_remove_paired_devices.ino [hunk 3]
```diff
@@ -1,91 +1,74 @@
...
-    Serial.println("No bonded device found.");
-  } else {
-    Serial.print("Bonded device count: "); Serial.println(count);
-    if(PAIR_MAX_DEVICES < count) {
-      count = PAIR_MAX_DEVICES; 
-      Serial.print("Reset bonded device count: "); Serial.println(count);
+  int count = SerialBT.getNumberOfBondedDevices();
+  if(!count){
+    Serial.println("No bonded devices found.");
+  }else{
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/examples/bt_remove_paired_devices/bt_remove_paired_devices.ino:46]:
```suggestion
  } else {
```

File: libraries/BluetoothSerial/examples/bt_remove_paired_devices/bt_remove_paired_devices.ino [hunk 4]
```diff
@@ -1,91 +1,74 @@
...
         }
-      }        
-    }
-  }
+        Serial.printf("Found bonded device #%d BDA:%s; Name:\"%s\"\n", i, bda2str(pairedDeviceBtAddr[i], bda_str, 18), rmt_name);
+        SerialBT.invalidateRemoteName(); // Allows waiting for next reading
+        if(REMOVE_BONDED_DEVICES){
+          if(SerialBT.deleteBondedDevice(pairedDeviceBtAddr[i])){
+            Serial.printf("Removed bonded device # %d\n", i);
+          }else{
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/examples/bt_remove_paired_devices/bt_remove_paired_devices.ino:65]:
```suggestion
          } else {
```

File: libraries/BluetoothSerial/examples/bt_remove_paired_devices/bt_remove_paired_devices.ino [hunk 5]
```diff
@@ -1,91 +1,74 @@
...
+          } // if(ESP_OK == tError)
+        } // if(REMOVE_BONDED_DEVICES)
+      } // for(int i = 0; i < count; i++)
+    } // if(ESP_OK == tError)
+  } // if(!count)
 }
- 
-void loop() {}
+
+void loop() {}
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/examples/bt_remove_paired_devices/bt_remove_paired_devices.ino:74]:
```suggestion
void loop() {}

```

Comment (lucasssvaz) [reply libraries/BluetoothSerial/examples/bt_remove_paired_devices/bt_remove_paired_devices.ino:74]: Keep the newline character so the source file has a posix compliant ending

File: libraries/BluetoothSerial/src/BluetoothSerial.cpp [hunk 6]
```diff
@@ -531,58 +539,90 @@ static void esp_bt_gap_cb(esp_bt_gap_cb_event_t event, esp_bt_gap_cb_param_t *pa
...
 
-        case ESP_BT_GAP_KEY_REQ_EVT:
+#ifdef CONFIG_BT_SSP_ENABLED
+        case ESP_BT_GAP_KEY_REQ_EVT: // Enum 8 - Security Simple Pairing Passkey request
             log_i("ESP_BT_GAP_KEY_REQ_EVT Please enter passkey!");
+            if (key_request_callback) {
+                memcpy(current_bd_addr, param->cfm_req.bda, sizeof(esp_bd_addr_t));
+                key_request_callback();
+            }
+            else {
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/src/BluetoothSerial.cpp:577]:
```suggestion
            } else {
```

File: libraries/BluetoothSerial/src/BluetoothSerial.cpp [hunk 7]
```diff
@@ -1211,4 +1288,108 @@ BTAddress BluetoothSerial::getBtAddressObject() {
...
+    if(!isReady(false, READY_TIMEOUT)){
+        log_w("Attempted to drop cache for uninitialized driver. First call begin()");
+        return;
+    }
+
+    int expected_dev_num = esp_bt_gap_get_bond_device_num();
+    if(expected_dev_num == 0){
+        log_i("No devices in cache.");
+        return;
+    }else{
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/src/BluetoothSerial.cpp:1358]:
```suggestion
    } else {
```

File: libraries/BluetoothSerial/src/BluetoothSerial.cpp [hunk 8]
```diff
@@ -1211,4 +1288,108 @@ BTAddress BluetoothSerial::getBtAddressObject() {
...
+    if(ret == ESP_OK){
+        if(dev_num != expected_dev_num){
+            log_w("Inconsistent number of bonded devices. Expected %d; returned %d",expected_dev_num, dev_num);
+        }
+        for(int i=0; i<dev_num; ++i){
+            ret = esp_bt_gap_remove_bond_device(dev_list[i]);
+            log_d("esp_bt_gap_remove_bond_device ret = %d", ret);
+            if(ret == ESP_OK){
+                log_d("Removed bonded device #%d", i);
+            }else{
```
Comment (lucasssvaz) [inline libraries/BluetoothSerial/src/BluetoothSerial.cpp:1385]:
```suggestion
            } else {
```

**Review by lucasssvaz — CHANGES_REQUESTED** [review]
The review is mostly nit-picks. I couldn't get the BT to work with the provided instructions. Please double check if it is working or my setup is faulty. Also squash the commits at the end, please.

File: libraries/BluetoothSerial/README.md [hunk 9]
```diff
@@ -1,19 +1,79 @@
...
 
-- Open the bluetooth terminal app
+- Open the Bluetooth terminal app and connect
 
 - Enjoy
+
+#### ESP32
+
+You can flash one of the ESP32 with the example [`SerialToSerialBTM`](https://github.com/espressif/arduino-esp32/blob/master/libraries/BluetoothSerial/examples/SerialToSerialBTM/SerialToSerialBTM.ino) (the Master) and another ESP32 with [`SerialToSerialBT`](https://github.com/espressif/arduino-esp32/blob/master/libraries/BluetoothSerial/examples/SerialToSerialBT/SerialToSerialBT.ino) (the Slave).
+Those examples are preset to work out-of-the-box but they should be scalable to connect multiple Slaves to the Master.
```
[the remark this answers is not in the document]
Comment (PilnyTomas) [reply libraries/BluetoothSerial/README.md:29]: Well, this is really strange... for me, it is also not working. However, I can connect to the slave from my smartphone.

[the remark this answers is not in the document]
Comment (lucasssvaz) [reply libraries/BluetoothSerial/README.md:29]: Yeah, using my smartphone I also managed to connect. There might be something weird happening in the master sketch.

**Comment by me-no-dev** [comment]
Closed in favor of https://github.com/espressif/arduino-esp32/pull/8723

**Pull request closed without merging**